fix: handle error on Close() of writable objects

aeijdenberg opened the pull request:
For objects (particularly files) that we write to, we should to check the return value of `f.Close()` and pass error back to the caller if this fails.

I recently read https://www.joeshaw.org/dont-defer-close-on-writable-files/ and while working on another PR in this repo I noticed that some some writable files had the `defer f.Close()` pattern, so I did a grep to find similar calls, and for cases involving writeable file, changed the defer function to capture the result of calling `Close()`, and if the `err` value of the parent is `nil`, then set it for the caller.

In this manner errors on close will be returned to the caller, but won't mask other errors if an error was already set.

This adjusts those to one of the patterns in that article, e.g.:

```golang
func doThing() (err error) {
	outFile, err := os.Create(dst)
	if err != nil {
		return err
	}

	// since we are writing to this file, make sure we handle errors
	defer func() {
		closeErr := outFile.Close()
		if err == nil {
			err = closeErr
		}
	}()

        .....
}
```

mtrmac reviewed (commented):
Thanks!

Just a quick note for now: This is definitely valuable; but I’ve been bitten by unintended named return value shadowing/overwriting several times in the past. Please use a clearly-unique name for the error value modified by `defer` (`retErr` is common in this codebase); for the other return values, it’s best not to add names to them them at all — use `_`.

Note to self: I didn’t actually review the changes.

aeijdenberg commented:
> Please use a clearly-unique name for the error value modified by `defer` (`retErr` is common in this codebase); for the other return values, it’s best not to add names to them them at all — use `_`.

Thanks, done as suggested, and squashed/rebased. I also removed the change to `pkg/blobcache/dest.go` - I'll have a go at that in a different PR as the entire method there looked to be quite unique in how it was doing its error handling flow.

mtrmac commented on storage/storage_dest.go:
```diff
@@ -272,7 +272,13 @@ func (s *storageImageDestination) putBlobToPendingFile(stream io.Reader, blobinf
+		if retErr == nil {
+			retErr = closeErr
+		}
+	}()
```
This function should exit before updating `s.lockProtected` and the like.

aeijdenberg replied:
Done.

mtrmac reviewed (commented):
Thanks!

In some of the situations, it seems inconsistent to recognize and report an error to write to a file, but to update records as if the file were written successfully anyway: the close + error detection should probably happen a bit earlier.

(Compare other code with “// A scope for defer” comments.)

aeijdenberg commented on ostree/ostree_dest.go:
```diff
@@ -148,8 +148,13 @@ func (d *ostreeImageDestination) PutBlobWithOptions(ctx context.Context, stream
 	}
-	defer blobFile.Close()
-
+	// since we are writing to this file, make sure we handle errors
```
Done.

aeijdenberg commented on pkg/blobinfocache/sqlite/sqlite.go:
```diff
@@ -82,12 +82,17 @@ func New(path string) (types.BlobInfoCache, error) {
 		return nil, fmt.Errorf("initializing blob info cache at %q: %w", path, err)
 	}
-	defer db.Close()
+	defer func() {
```
Done.

mtrmac commented on storage/storage_dest.go:
```diff
@@ -257,7 +257,7 @@ func (s *storageImageDestination) PutBlobWithOptions(ctx context.Context, stream
 // putBlobToPendingFile implements ImageDestination.PutBlobWithOptions, storing stream into an on-disk file.
 // The caller must arrange the blob to be eventually committed using s.commitLayer().
-func (s *storageImageDestination) putBlobToPendingFile(stream io.Reader, blobinfo types.BlobInfo, options *private.PutBlobOptions) (private.UploadedBlob, error) {
+func (s *storageImageDestination) putBlobToPendingFile(stream io.Reader, blobinfo types.BlobInfo, options *private.PutBlobOptions) (_ private.UploadedBlob, retErr error) {
```
I think this `retErr` is now unused (and shadowed by the nested one) and can be dropped.

aeijdenberg replied:
Thanks - I missed that one. Fixed.

mtrmac reviewed (commented):
Thanks! LGTM otherwise.

mtrmac commented:
Can you rebase, please?

mtrmac commented:
Thanks again!
